Add first version of the idr0021 segmentation notebook

dominikl opened the pull request:
Adds a notebook which performs the  idr0021 segmentation completely in R (to be used for the Cambridge workshop 6th Dec).

Test on : https://ome-lochy.openmicroscopy.org/jupyterhub-internal  (training-notebooks:master_merge_daily)

dominikl committed:
Add first version of the idr0021 segmentation notebook

dominikl commented:
/cc @pwalczysko @jburel

jburel commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,605 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Load the necessary [romero-gateway](https://github.com/ome/rOMERO-gateway) and [EBImage](http://bioconductor.org/packages/release/bioc/html/EBImage.html)EBImage library:"
```
``https`` for bioconductor link can be used

jburel replied:
Add a description to the notebook indicating its purpose
Also mention the version used

jburel commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,605 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "pixels <- getPixelValues(image, 1, 1, 2)\n",
```
explain what the numbers are

jburel commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,605 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise:** Play around with the parameters of the thresh function to get a feeling for how the parameters affect the segmentation, or even plug in other methods into the segmentation workflow to improve the result, see [EBImage documentation](https://www.rdocumentation.org/packages/EBImage/versions/4.14.2)"
```
Replace Play around by modify

jburel commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,605 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "We can create an Ellipse for each object taking the x,y coordinates (m.cx and m.cy), the radii (this is the m.majoraxis and calculate the 'minor' axis from 'major' axis and the eccentricity (see documentation of the computeFeatures method)) and the angle theta by which the Ellipse might be rotated."
```
computeFeatures  link?

jburel commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,605 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The dataframe is now attached to the image as HDF table. You can download and open it with software like HDF Compass, load it into python scripts using 'h5py' library, etc. or simply load it directly as R dataframe again."
```
an HDF

jburel commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,605 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Alternatively attach it as CSV file:"
```
a CSV file

pwalczysko replied:
I would also replace ``it``. I suppose it shoiuld be ``the dataframe``

jburel commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,605 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise**: Find the attached CSV file and open it in Excel"
```
or a software of your choice

jburel commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,605 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Task: Segment each image of a dataset, create an ROI for each of the objects, and summarize the results as dataframe, only taking the features of the largest ROI of an image into account.\n",
```
Exercise instead of Task

pwalczysko replied:
``...as dataframe, only taking the features of the largest ROI of an image into account.`` ->  This tries to say too much in one sentence. 

1. You are not summarizing the ROI features (the ROIs are just ellipses)
2. You are summarizing the dataframe, where the values come directly from the segmentation, which is more precise than the ROIs
3. The selection of the largest object cannot be thus defined with the word ``ROI`` - see points 1. and 2 above

dominikl committed:
Add more text/explanations to the notebook; removed IDs

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "source": [
+    "# Image Segmentation using the idr0021 dataset\n",
+    "\n",
+    "This notebook uses the idr0021 dataset (original: [idr0021-lawo-pericentriolarmaterial/experimentA](http://idr.openmicroscopy.org/webclient/?show=project-51)) and tries to reproduce some of the analysis published in ['Subdiffraction imaging of centrosomes reveals higher-order organizational features of pericentriolar material'](https://doi.org/10.1038/ncb2591). In particular we're trying to  ...
```
Leave the word ``original`` out - it is not precise enough.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+    "\n",
+    "This notebook uses the idr0021 dataset (original: [idr0021-lawo-pericentriolarmaterial/experimentA](http://idr.openmicroscopy.org/webclient/?show=project-51)) and tries to reproduce some of the analysis published in ['Subdiffraction imaging of centrosomes reveals higher-order organizational features of pericentriolar material'](https://doi.org/10.1038/ncb2591). In particular we're trying to  ...
+    "\n",
+    "Each user has a partial copy of the dataset ('Lab1' / YOUR USER / Project: 'idr0021'). Some of the images are also part of the dataset 'R-dataset' which you can use for practicing the segmentation and annotation steps."
```
Instead of ``('Lab1' / YOUR USER / Project: 'idr0021')`` just ``...inside the Project "idr0021"

pwalczysko replied:
But actually, this is so very much outreach specific... Maybe we would like to put such sentence rather in the workflow .pdf than here.

jburel replied:
Agree better to put in the pdf

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "metadata": {},
+   "source": [
+    "The grey blocks with the In\\[ \\] notation to the left are code blocks. Click on a block to select it and click the play button in the toolbar to execute the block of code (alternatively hold SHIFT key and hit RETURN). The notation will change to In\\[*\\] to indicate that this code is currently running. The output is displayed below the block. Wait for it to finish. When the execution is f ...
+    "\n",
```
RETURN -> ENTER

pwalczysko replied:
I would not talk about notations. 
``The blocks with the sqare brackets to the left of them...`` seems to me more straightforward.


``The notation will change...`` sounds strange. 
Maybe: ``The star in the square brackets to the left of the cell will turn into a number.``

pwalczysko replied:
For some reason, what should be ``[*]`` here, resolves as ``ln[*]`` in the notebook. What does the ``ln`` mean ? I think it should be removed.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+    "message(\"Just wait for a bit...\")\n",
+    "Sys.sleep(3)\n",
+    "message(\"Done.\")"
+   ]
```
Nice exercise

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Load the necessary [romero-gateway](https://github.com/ome/rOMERO-gateway) and [EBImage](https://bioconductor.org/packages/release/bioc/html/EBImage.html) library. We're currently using romero-gateway 0.4.3, EBImage 4.13.0 with R 3.4.1 . "
```
``Load the necessary...`` -> ``romero-gateway and EBImage libraries are essential for the R <-> OMERO communication. We will load the libraries in this cell.``

pwalczysko replied:
``We're`` -> ``We are``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Load and display an image"
```
Rather: ``Load an image from OMERO.server and display it in R``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Load an image from the example dataset 'R-dataset' (these images are part of the original 'CDK5RAP2-C' dataset).\n",
```
``these images are parto of the original...dataset`` -> Leave this bit completely out, it is very hard to understand and not really necessary for the moment to deal with. 
I think that if we want to be precise and explain the similarities and differences between the R-dataset and the idr0021 datasets, we could do it in the workflow .pdf.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "source": [
+    "Load an image from the example dataset 'R-dataset' (these images are part of the original 'CDK5RAP2-C' dataset).\n",
+    "\n",
+    "**Exercise:** Go to the image 'siControl_N20_Cep215_I_20110411_Mon-1509_0_SIR_PRJ.dv' in your 'R-dataset', find the 'Image ID' and copy/paste it below. You can use a different image of the dataset if you like, but I recommend that one because it nicely shows the toroid centrioles."
```
``I`` -> ``We`` :)

But maybe rather:

``Select the image 'siControl_N20_Cep215_I_20110411_Mon-1509_0_SIR_PRJ.dv' in the "R-dataset" in OMERO. From the right-hand pane, copy the ImageID. Paste the ID into the code cell below.``

pwalczysko replied:
Leave out the sentence aoubt ``You can use a different image if you like....`` It is too long. Nobody knows what the centrioles are at this stage. I guess they will understand that they can use another image. 

Possibly additional exercies:


``Exercise: Find in OMERO.web an ImageID of a different image and display it in R.``

pwalczysko replied:
```
Found more than one class "Image" in cache; using the first, from namespace 'romero.gateway'
Also defined by ‘EBImage’

```

This is the output. Hum, not really nice.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Log into the OMERO server"
```
``Log into`` -> ``Log in to``. 

Maybe also add the explanation about

``Execute this cell first, this will prompt you to enter your credentials.``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise:** Modify the parameters of the thresh function to get a feeling for how the parameters affect the segmentation, or even plug in other methods into the segmentation workflow to improve the result, see [EBImage documentation](https://www.rdocumentation.org/packages/EBImage/versions/4.14.2)"
```
``....or even plug in...`` Leave this part of the sentence out. If you want to, declare another (more advanced) exercise instead. Going to the unfamiliar doc and looking for an entity which I do not fully understand is suddenly quite a jump.

pwalczysko replied:
After I have executed this cell, I am getting a very unpleasant effect - not seeing the resulting image without scrolling.

![screen shot 2018-11-21 at 12 09 03](https://user-images.githubusercontent.com/2478303/48840378-5870fb00-ed86-11e8-8565-46fd9ec998ee.png)

pwalczysko replied:
It is because of the width and height parameters - this is not explained in the exercise. The starting parameters for w and h should be something like 
``threshImg <- thresh(ebi, w=15, h=40, offset=0.1)``
which for me displays the whole image at least.

pwalczysko replied:
This exercise is simply confusing. 

When I hear ``thresh`` I immediately think about ``thresholding``. The problem is that in the first line, I am actually manipulating the size of the output -not at all the usual thresholding stuff. The function is not explained. Neither it is intuitive. And the link to the doc is far too general (send me directly to the explanation of the ``thresh`` function please.

pwalczysko replied:
It is the ``offset`` parameter which really does something more intuitive. But, alas, it has not a very intuitive name...

Also, you did not explain what the objects to be thresholded for in the image are...

pwalczysko replied:
You need to offer this link https://www.rdocumentation.org/packages/EBImage/versions/4.14.2/topics/thresh - and quite soon after the start of the cell/exercise.

dominikl replied:
w, h and offset are thresholding parameters, nothing to do with the image size of the output. That the image size changed, must have been some weird random artifact. But yes, a link to the documentation of the function is definitely needed earlier.

pwalczysko replied:
w and h are "width and height of the moving rectangular window". Which is not very intuitive. In fact still have no idea what is it doing. Yes, you are right, it is not resizing the output window. But neither it is intuitive, nor it intuitively belongs to a thresholding function.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Compute the features (measurements)"
```
Oh, this is not explained at all - what features am I computing ? What functions am I using ?

dominikl replied:
Well, the functions in the code block. And the features or measuerments are displayed as table when the code block ran. Does that really need more explanation? Can add a link to the computeFeatures documentation maybe.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "metadata": {},
+   "source": [
+    "We can create an Ellipse for each object taking the x,y coordinates (m.cx and m.cy), the radii (this is the m.majoraxis and calculate the 'minor' axis from 'major' axis and the eccentricity (see [documentation](https://www.rdocumentation.org/packages/EBImage/versions/4.14.2/topics/computeFeatures) of the computeFeatures method)) and the angle theta by which the Ellipse might be rotated.\n",
+    "\n",
```
First sentence:

``We will create an Ellipse for each object.``
Second:

``We will take the coordinates (from where ?) , the radii, the eccentricity and the angle theta from... (the table produce by the previous exercise ?).``

Third:

``All the rest about how one radius is counted from another, about the link to docs (more precise link please)...``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise:** Open the image in OMERO.web and compare the ROIs with the segmented image."
```
I would not compare the ROIs with the segmented image. The segmented image is somewhere else (in the Jupyter notebook). Also, why not to use the OMERO.iviewer itself, the image is there with the ROIs on it, so I can see immediately the quality of my ROIs ?

When I opened my ROIs in iviewer, I noticed they are not associated with any channel. Is that a feature you are using later ?

dominikl replied:
That's what I actually meant, just open the image in OMERO.web (iviewer) and have a look at the ROIs. 

Just realized I've not implemented the 'setChannel' functionality in the R gateway yet, so I can't associate the channel with ROI yet.

pwalczysko replied:
>That's what I actually meant, just open the image in OMERO.web (iviewer) and have a look at the ROIs.

@dominikl If this is so, then just remove the words ``segmented image``. The image in iviewer will not be segmented.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Attach the whole dataframe to the image:"
```
I suppose the dataframe is called ``features`` ? I would write it here, it is quite far up in the code to find out.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "invisible(attachDataframe(image, features, \"ROI features\")) # the invisible just suppresses some unneccessary output"
```
the function ``attachDataframe`` is very magical, because it was not mentioned before. @dominikl is telling me that it is an romero-gateway method. But we do not have any link to the doc (the only doc we could generate would be a local html - what would be the best compromise here @jburel  ?

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "source": [
+    "Task: Segment each image of a dataset, create an ROI for each of the objects, and summarize the results as dataframe, only taking the features of the largest ROI of an image into account.\n",
+    "\n",
+    "Note: We need to be able to specify the channel which is analyzed by its name.\n",
```
``We need to be able to specify the channel which is analyzed by its name.`` -> ``Only one channel in each image is relevant for our analysis. The relevant channel is identified by its name.``

this would be re-written what you have there - but didn't you also want to say that the name of the channel should match the name of the dataset containing the image ?

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+    "\n",
+    "Note: We need to be able to specify the channel which is analyzed by its name.\n",
+    "\n",
+    "We put the all the pieces together and wrap them up in a function:"
```
``...the all the pieces ...`` -> to many ``the`` here. 

I would be more specific aoout what the ``pieces`` are.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise:** Check the ROIs in OMERO.web"
```
I would put this ``Exercise`` above the code block. Then, I would reformulate the exercise 
``Insert the Dataset ID of your "R-dataset" in OMERO into the code below. Make sure you are selecting the Dataset in the left-hand pane of OMERO.web, not an image, when retrieving the Dataset ID. Execute the code cell. Then, check the ROIs in OMERO.web.``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise:** The measurements (Area, Perimeter and Diameter) are currently showing pixels. Find out the pixel size in nanometer and apply it accordingly."
```
Where should I find the pixel size in nanometers ? In OMERO.web ? Write it please.

``Apply it accordingly...`` -> Do you mean I should then find a place in the code where the calibration is being applied and change it, then rerun the cell ?

pwalczysko replied:
Discussed with @dominikl : 

1. highlight this Exercie as (advanced)
2. formulate something like ``add into the cell a code snippet which transforms the pixel values to nanometers. Then rerun the cell.``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Do not run this, FYI only!**\n",
```
Put this information higher (above the previous header). Remove the word ``run``. Maybe ``For your information only, do not execute``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "source": [
+    "**Do not run this, FYI only!**\n",
+    "\n",
+    "The next code block does the same what we did before but across the whole idr0021 project (taking into account that the channel name which has to be analyzed is equal to the dataset name). This can take a while, therefore we skip this step and load the results from the attached dataframe, see next step."
```
``... does the same what we did before...`` -> ``...reuses the code from previous exercises and applies them on the whole idr0021 project (over 400 images in 10 Datasets)``

``....but...`` Make a new sentence, do not use ``but``

pwalczysko replied:
``(taking into account that the channel name which has to be analyzed is equal to the dataset name)``
Make a new sentence out of this. See above my suggestion about the ``relevant`` channel formulations

pwalczysko replied:
As discussed, it would be good to add a csv as well as the OMERO.table to the project.

pwalczysko commented:
The idr0021 script here (the one which is FYI only for participants) is working fine -> tried the whole run on user-40, worked as expected.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Plot the data"
```
The plot y-axis shoiuld have more ticks, maybe every 100 or so.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise:** Find the outlier in OMERO.web (using 'parade') and check what might have gone wrong there."
```
``outlier`` -> ``outliers`` 

OMERO.parade will be introduced only later, after this. But, we can highlight this exercise as one to do after the parade is explained.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### Compute some statistics"
```
``Compute`` -> ``Calculate``

Remove ``some`` - sounds too offhand I guess

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,702 @@
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
```
very little output here, very little description. Is that the t-test ? or what is acutally going on here ? I guess there was a better explanation in the old R notebook ?

pwalczysko commented:
Finished reading. Several comments, but looks pretty good in general.

dominikl committed:
Fixed issues

dominikl commented:
Hope I fixed all your remarks @pwalczysko . Removed the "trainer only" part from the notebook. That was the part which runs the segmentation over the whole idr0021 project in order to generate the trainer-1 'Summary from R' table. Can add that as seperate notebook, if needed. Also added "Task" section in the description of the notebook which outlines each task which is demonstrated by the notebook. But I'm still not sure if it is generic enough to be used as general ITR example notebook, the "Exercise" parts are certainly not.

I still need to add the 'Summary from R' as CSV file to the trainer-1 project, and add another exercise "Download the CSV and open in Excel" at the end of this notebook, but that will be just a small addition.

dominikl commented:
From the last testing round I got:

- Use a 'run' flag to en/disable optional exercise block (pixel size exercise)
- Add solution for it to bottom
- Make clear to load the trainer-1 idr0021 project for stats part
- If there's no output, print out at least a message like "Dataframe has been attached".

I'll push another commit addressing these issues later. More feedback welcome. Will try to finalize the notebook until Monday.

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,677 @@
+    "\n",
+    "- Segment each image of a dataset.\n",
+    "- For visualisation create an ROI for each of the objects detected.\n",
+    "- For further analysis only take the largest object per image into account (assumption: this is a centriol), in particular it's area, perimeter and diameter.\n",
```
``centriol`` -> ``centriole``   (missing ``e``)
``in particular it's area, perimeter and diameter`` -> ``Parameters to evaluate for this object are area, perimeter and diameter.`` (it is hard to understand what ``it's`` is in the sentence. If there are more parameters than those three, just write ``Main parameters...``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,677 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise:** Go to the image 'siControl_N20_Cep215_I_20110411_Mon-1509_0_SIR_PRJ.dv' in your 'R-dataset', find the 'Image ID' and copy/paste it below."
```
``copy/paste it...`` -> ``copy it and paste it...``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,677 @@
+    "\n",
+    "Note: Up to now we still have a binary image (the pixel array only consists of 0's (background) and 1's (objects), therefore we need:\n",
+    "\n",
+    "- [bwlabel](https://www.rdocumentation.org/packages/EBImage/versions/4.14.2/topics/bwlabel) to distinguish the single objects from each other. It assigns a different number to each object (first object will get pixel values '1', the second pixel values '2', etc.)"
```
great explanation

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,677 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "For the further analysis we need the object sizes: 's.area', 's.perimeter', especially 's.radius.mean' (~ diameter).\n",
```
``For the further analysis...`` -> ``For further analysis...``

``...object sizes`` -> ``...object metrics describing the size:``


``, especially...`` -> ``...and most importantly the...``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,677 @@
+    "Alternatively attach it as CSV file:\n",
+    "\n",
+    "- Use R's built-in write.csv method to create a csv file\n",
+    "- And attach that to the image using attachFile function from the romero.gateway"
```
I have in my csv only 2 lines, but I am having 7 ROIs in that image. Do you pre-select the objects to be evaluated ?

pwalczysko replied:
Deleting the comment about the csv having only 2 lines - it is invalid, sorry, the number of lines is 7 as it should be

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,677 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Exercise:** Open some of the other images of the dataset in the full viewer to check the ROIs."
```
In the table above this exercise, what is ``ROI Index`` ? Number of ROIs ? If so, could in be called ROI number ?

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,677 @@
+    "## Statistical analysis\n",
+    "\n",
+    "The image segmentation has already been run over the whole idr0021 project and the results attached to the project as table 'Summary from R'. \n",
+    "Instead of running the segmentation repeatedly we just load the table as dataframe 'centrioles':"
```
``Instead of running the segmentation repeatedly we just load the table as dataframe 'centrioles' `` ->

``We are going to load this table from OMERO as dataframe "centrioles"``

pwalczysko replied:
Maybe add: ``In this manner, we will quickly obtain a sizeable segmentation dataset to perform a statistical analysis on.``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,677 @@
+   "outputs": [],
+   "source": [
+    "# 'attachDataframe' directly attaches an R dataframe to an OMERO image (project, dataset, etc.)\n",
+    "# ('invisible' not strictly necessary, just suppresses some unneeded output)\n",
```
unneeded -> unnecessary  (as @manics puts it, ``unnecessary sounds slightly better``)  :)

pwalczysko commented:
Couple of more edits, mainly text improvements.

@dominikl Regarding the scripts for cleanups.

We have scripts which delete:

1. All the ROIs in a specific dataset - this works very well for me https://github.com/ome/training-scripts/blob/master/maintenance/scripts/delete_ROIs.py
2. All annotations (also attachments) on **images** in a specific dataset  https://github.com/ome/training-scripts/blob/master/maintenance/scripts/delete_annotations.py - needs improving, does not iterate through all users

We do not have a script which would delete:
1. All annotations on all specified **datasets** (not images inside those datasets). This should be probably an additional option of https://github.com/ome/training-scripts/blob/master/maintenance/scripts/delete_annotations.py - that obviously needs improvement

dominikl committed:
Some more corrections

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,717 @@
+    "- Filtering to remove noise, using [medianFilter](https://www.rdocumentation.org/packages/EBImage/versions/4.14.2/topics/medianFilter).\n",
+    "- [fillHull](https://www.rdocumentation.org/packages/EBImage/versions/4.14.2/topics/fillHull) to close holes in the segmented objects.\n",
+    "\n",
+    "Note: Up to now we still have a binary image (the pixel array only consists of 0's (background) and 1's (objects), therefore we need:\n",
```
``Up to now...`` -> ``Up till now...`` (alternative: ``Until now...``) - approved by @mtbc

pwalczysko replied:
I guess, also ``have`` -> ``had`` (because ``Up till now...`` is in the past)

pwalczysko replied:
The bracket starting at ``(the pixel array...`` does not seem to be closed later

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,717 @@
+    "- For visualisation create an ROI for each of the objects detected.\n",
+    "- For further analysis only take the largest object per image into account (assumption: this is a centriole). Parameters to evaluate for this object are area, perimeter and diameter\n",
+    "\n",
+    "Note: We need to be able to specify the channel which is analyzed by its name. Only one channel in each image is relevant for our analysis. The relevant channel is identified by its name.\n",
```
I would simply remove the words ``...which is analyzed by its name.`` You are explaining it well enough in the following sentences, and with better wording even.

pwalczysko replied:
Add a sentence:

``The name must match the name of the containing dataset.``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,717 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Iterate over the dataset and call run the analyze function for each of the images:"
```
Ether ``call`` or ``run``, not both. I would say ``call``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,717 @@
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "**Optional Exercise:** The measurements (Area, Perimeter and Diameter) are currently showing the number of pixels. Find out the pixel size in nanometer and apply it accordingly. Hint: The pixel size is the same for all images. Select one image in OMERO.web and check its metadata (right hand side panel). Solution: See bottom."
```
``nanometer`` -> ``nanometers`` (should be plural)

pwalczysko replied:
``Solution: See bottom`` -> ``Solution: See the bottom of this notebook.``

pwalczysko replied:
``Hint: The pixel size...``  -> ``Hint: The X and Y pixel sizes ...``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,717 @@
+    "    \n",
+    "    # Just print out the data frame again to check result:\n",
+    "    result\n",
+    "}"
```
```
Error in parse(text = x, srcfile = src): <text>:11:1: unexpected '}'
10:     result
11: }
    ^
Traceback:
```

That is what I get if I replace nothing and just execute the cell as it is.
Was that the idea ? I thought it would smoothly pass saying "whatever, you did not bother to do the exercise", instead of erroring ?

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,717 @@
+   "outputs": [],
+   "source": [
+    "pxSizeInNM <- NA # REPLACE_WITH_PIXEL_SIZE\n",
+    "if (!is.na(pxSizeInNM))\n",
```
Opening curly braces missing for ``if``

pwalczysko commented on R/idr0021 - Segmentation.ipynb:
```diff
@@ -0,0 +1,717 @@
+    "  testResults <- rbind(testResults, data.frame(Protein_1=param[1], Protein_2=param[2], p_value=pval))\n",
+    "}\n",
+    "testResults"
+   ]
```
After running this cell, I get

```
Warning message in wilcox.test.default(x = c(211.522222470665, 214.349311216569, :
“cannot compute exact p-value with ties”
```

Yes, not too worrying, but of course people will ask what that is ?

dominikl replied:
The warning basically means that there are two values which are exactly the same ("tie"). I'll have a look and hopefully find an explanation which I can use during the workshop.

dominikl replied:
Yes, coincidentaly there are really two ROIs which are exactly(!) the same size and diameter, amazing.

pwalczysko commented:
Again, just couple of minor comments, the notebook reads and runs well.

jburel commented:
@dominikl Could you fix the minor comments this am? we will need to have that PR merged 
so we can have the image built and deployed for Monday testing

dominikl committed:
More minor corrections

dominikl commented:
Ah sorry, forgot to add/push the changes yesterday. Done now.

jburel commented:
Thanks so We will need to trigger the build for a final check

pwalczysko commented:
Happy with the changes
Ready to merge

jburel commented:
Thanks 
Merging
